### ISSUE TITLE
how to init dds in the process which is a child process forked out

### ISSUE THREAD

**Issue by sgf201** [eclipse-cyclonedds/cyclonedds#1105]
I am trying to use cyclonedds  to replace another comm component ,the target APP 's fixed flow is：

main（）
{
     init dds
     do something;
     dds read  or write;
     if(xxx)
     {
         if(fork()==0)
         {
              init dds;
              while(1)
              do something
              dds read or write;
         }
         else
         {
              ;;
         }
     }
     
}
 

so, I need to init dds in child process,I try  like this:

main () {
create_participant
create_readet0;
  if (fork () == 0)
  create_reader1
  else
  create_reader2
}

I test the situation above , the reader1 can not get message, but  all the cyclonedds apis return no err

main () {
create_participant
if (fork () == 0)
{
create_participant
create_reader1
}
else
create_reader2
}

I test the situation above , the reader1 can not get message ether. AND, all the cyclonedds apis return no err

main () {
create_participant
if (fork () == 0)
{
dds_delete(participant)
create_participant
create_reader1
}
else
while(1);
}

I test the situation above , the reader1 process crashed ,so sad.

father process  init DDS and use it  before fork a child ,and I think it is normal enough
So,   how could I init DDS in child process ????
@eboasson 
@rotu 
@lygstate 
@kartben 
@clalancette

**Comment by eboasson** [comment]
@sgf201 I don't understand why you are pinging various people who aren't involved in this (nor why you didn't simply reopen #1097 after GitHub closed it automatically because I merged #1103). But whatever.

I have been giving it some thought. Would you be good enough if the child leaked some of the memory allocated by Cyclone in the parent? Because then I think it should not be very hard to re-initialize Cyclone in the child. It is going to be very difficult to do so without any memory leaks because, if that is at all possible.

**Comment by sgf201** [comment]
> @sgf201 I don't understand why you are pinging various people who aren't involved in this (nor why you didn't simply reopen #1097 after GitHub closed it automatically because I merged #1103). But whatever.
> 
> I have been giving it some thought. Would you be good enough if the child leaked some of the memory allocated by Cyclone in the parent? Because then I think it should not be very hard to re-initialize Cyclone in the child. It is going to be very difficult to do so without any memory leaks because, if that is at all possible.

sorry for all, I break down a little at the task of DDS apply. I have spend much more time at debug. And it is really do not work.

next, a little explanation,
(nor why you didn't simply reopen #1097 after GitHub closed it automatically because I merged #1103):
I thought that you define #1103 is a total separate DDS init,and this one is a abnormal usage problem. so I write it in a new issue.

 why you are pinging various people who aren't involved in this:
 I'm kind of in a hurry to sort it out，I really delayed a lot in debug. And I am not sure whether you are busy or think that is so useless to spend time on it, so a day after Posting, I added others. any one's reply is precious. I need to get a result as soon as possible.

at last, the problem,
I do not care memory leaks at this moment, It's great if it just works.
I know that "you think it should not be very hard to re-initialize Cyclone in the child", because I used to think that way either, but cycloneDDS really  do not work in the child process which is forked out, reinit  directly  or delete old &init new, do not work at all, and crash in some cases.

any method to use cycloneDDS is  precious in the fixed app situation 

look forward to your kind advice
@eboasson

**Comment by eboasson** [comment]
@sgf201 try this: https://github.com/eboasson/cyclonedds/commit/eff7e8208ea5c02dda72f206e5561748961b8334 If you call this new `dds_ruthless_reset_for_child_after_fork` function in the child after forking, it resets everything by just `memset`ting things. It worked for a simple experiment I did, which was also `fork`ing somewhere in the "hello world" subscriber.

YMMV; obviously this needs a bit more care and, probably, a different name. I do like to know whether it addresses your problem. If it does, we'll clean it up and in the mean time you can make progress. If it doesn't, I think it must mean I overlooked something else I should have reset, so we just do another round.

**Comment by sgf201** [comment]
> @sgf201 try this: [eboasson@eff7e82](https://github.com/eboasson/cyclonedds/commit/eff7e8208ea5c02dda72f206e5561748961b8334) If you call this new `dds_ruthless_reset_for_child_after_fork` function in the child after forking, it resets everything by just `memset`ting things. It worked for a simple experiment I did, which was also `fork`ing somewhere in the "hello world" subscriber.
> 
> YMMV; obviously this needs a bit more care and, probably, a different name. I do like to know whether it addresses your problem. If it does, we'll clean it up and in the mean time you can make progress. If it doesn't, I think it must mean I overlooked something else I should have reset, so we just do another round.

the patch works, thanks a lot~
And, after this work around, I`d like to discuss whether the fixed flow above like will be supported in the future, I is important for us to make final decision about middleware choice

**Comment by eboasson** [comment]
Doing nothing means you, or anyone else in a similar situation, can't use Cyclone. This is such a simple solution that I don't see a problem supporting it. The documentation would have to include a statement that it may leak memory allocated in the parent and that it must not ever be called in parallel to any other operation, but I'd be ok with that. Making it a supported interface means adding a test, but that is also easy.

So the only hard part is coming up with a decent name ... Something like `dds_pthread_atfork_child_handler` could work, I suppose.

**Comment by sgf201** [comment]
![image](https://user-images.githubusercontent.com/5253415/157832026-7b6fd650-3ca4-4ee7-aab9-1b734d906e12.png)
after apply this: [eboasson@eff7e82](https://github.com/eboasson/cyclonedds/commit/eff7e8208ea5c02dda72f206e5561748961b8334) , the father process and child process raise the same err,but it seems not fatal. some data lost which from another host.

**Comment by elfenpiff** [comment]
@eboasson This was also raised as an issue in iceoryx, here are some thoughts from our side.

From a first look everything seems to be in order in our code base but this piece of our code is not as well tested as it should be.
The error call will only be raised when calling `iox_sub_release_chunk` on an already released chunk.
